### ISSUE TITLE
Replace cabal project parsing with Parsec

### ISSUE THREAD

**Issue by typedrat** [haskell/cabal#6101]
I plan on doing this, since as it stands I don't see a way to implement the sort of overriding syntax needed for #5522, and even if it is possible we should eat our veggies.

**Comment by phadej** [comment]
@typedrat if you can pull it off, it would be super super great!

**Comment by Cmdv** [comment]
Is the idea here to use the internal Parsec representation `/cabal/Cabal-syntax/src/Distribution/Parsec/`?
If so where should one start looking for `project parsing`?

**Comment by ulysses4ever** [comment]
> Is the idea here to use the internal Parsec representation /cabal/Cabal-syntax/src/Distribution/Parsec/?

Yes.

> If so where should one start looking for project parsing?

I was wondering this myself, and it was tricky. At this point, rather than waiting for someone knowledgeable to reply, I'd compile it with profiling on, run it with -xc on a buggy project file, see it failing and study the stack trace...

**Comment by ulysses4ever** [comment]
>  where should one start looking for project parsing?

Less discouraging answer is: mostly, https://github.com/haskell/cabal/blob/master/cabal-install/src/Distribution/Client/ParseUtils.hs

**Comment by Cmdv** [comment]
@ulysses4ever thanks I took a look prior to your comment but never came back to ask about what the "buggy project" was, as the linked issue was a little above my head right now 😓 

if you have a small example of what I'd need to run to see this error that would be great, in the mean time I'll also look at the linked file you've given 👍

**Comment by ulysses4ever** [comment]
By buggy I meant a project file with a parse error. This should throw an error from the parser and inspecting the stack trace from the error should bring some insights on where the parser is located and how it's called.

I'll see about an example but don't hold your breath, please.

**Comment by Cmdv** [comment]
ah ok I'm with you not to worry cheers for the help... I'll report back if I start this and if not then someone else can take a look.

**Comment by ulysses4ever** [comment]
@jgotoh I can confirm that all that your findings make sense to me. Godspeed with your experiment. Do ask questions when stuck and try to post a draft PR early — even if it doesn't do much yet.

**Comment by Cmdv** [comment]
@jgotoh you are right I wasn't able to get time to go too far into this!! looks like you've done some very good investigative work 👌 I'll pass the torch over to you (though I only ran a couple a metres 😂)

**Comment by jgotoh** [comment]
Many thanks to both of you! :) I will keep you updated and will try to create some rough draft soon to gather some feedback. Most updates will happen on weekends though because of work.
